Remove use of HAS_GPU from `dispatch` functions

oliverholworthy opened the pull request:
Remove use of HAS_GPU from `dispatch` functions. 

Following #211 the variable `HAS_GPU` no longer includes the information about whether or not cudf is installed. This can result in confusing errors when using the dispatch functions in a GPU enabled environment without cudf available. (`'NoneType' object has no attribute 'DataFrame'`)

karlhigley commented on merlin/core/dispatch.py:
```diff
@@ -24,16 +24,16 @@
 rmm = None
 
-if HAS_GPU:
+if cudf:
```
I could be missing something, but seems like `cudf` would need to imported from `merlin.core.compat` for this to work

oliverholworthy replied:
I missed this commit from moving from the other PR. Updated

oliverholworthy commented on merlin/core/compat.py:
```diff
@@ -61,6 +61,11 @@ def _get_gpu_count():
     cupy = None
 
+try:
+    import cudf
```
Based on your comment here ("[may blow up in CPU-only environments](url)") @karlhigley  

This is currently the case if you try to import cudf in our containers without GPUs available, you'll get a Segmentation fault, which the except ImportError can't catch

```
Exception ignored in: 'cuda._lib.ccudart.utils.cudaPythonGlobal.lazyInitGlobal'
Traceback (most recent call last):
  File "cuda/_cuda/ccuda.pyx", line 3671, in cuda._cuda.ccuda._cuInit
  File "cuda/_cuda/ccuda.pyx", line 435, in cuda._cuda.ccuda.cuPythonInit
RuntimeError: Failed to dlopen libcuda.so
Segmentation fault (core dumped)
```

oliverholworthy replied:
The segmentation fault was fixed in cuda-python 11.7.1 , however that requires use of cudf 22.12 and above. With that in mind I think it might be ok to keep the import as it is and say that support for CPU-only environments that have cudf installed requries cudf >=22.12 ?

karlhigley replied:
Yeah, that sounds good. You're way ahead of me (as usual.)

karlhigley commented:
It looks like some of the NVT test failures might actually be related to these changes

oliverholworthy commented:
> It looks like some of the NVT test failures might actually be related to these changes

I've updated and resolved these issues. The typehint for `read_dispatch` suggested that it might only accept a dataframe type, but is also called with strings. Updated to handle that case. 

And the functions `{random_state, arange, array, zeros}` all had the same pattern. Updated to handle RangeIndex type passed as `like_df` and the base case where like_df is None and we're simply picking between cupy and numpy. The code is more verbose, but is clearer about what it's doing now. Hopefully a further refactor later will remove the need for these altogether
